Get serial number for astromechs and use dev_node_pattern

JAAlvarado-Montes opened the pull request:
Add code to get the serial number of astromechanics controllers. Also, add the `dev_node_pattern` parameter.

## Description
Get the serial number of astromechanics controllers and add the `dev_node_pattern` parameter so it can be used to search for both astromechanics and birger controllers and initialize them.

## How Has This Been Tested?
The changes were tested locally using real hardware.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -8,6 +15,13 @@ class Focuser(AbstractSerialFocuser):
         model (str, optional): default 'Canon EF/EF-S'
+        initial_position (int, optional): if given the focuser will drive to this encoder position
+            following initialisation.
+        dev_node_pattern (str, optional): Unix shell pattern to use to identify device nodes that
+            may have an Astromechanics adaptor attached. Default is '/dev/tty.usbserial-AG0*.?',
+            which is intended to match all the nodes created by FT232R USB UART serial
+            adaptors, as used at the time of writing by Huntsman.
```
This should default to `None` in POCS and then be set via the config.

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -20,8 +34,15 @@ class Focuser(AbstractSerialFocuser):
                  name='Astromechanics Focuser Controller',
                  model='astromechanics',
+                 initial_position=0,
+                 dev_node_pattern='/dev/tty.usbserial-AG0*',
+                 serial_number_pattern=serial_number_pattern,
```
The dev node pattern should be `None` in POCS to keep it generic and the values can come from the config.

But, if you are not doing anything with these parameters here and they are already in the super class, you don't really need to specify them, they will just go through with `**kwargs`

```suggestion
                 dev_node_pattern=None,
                 serial_number_pattern=serial_number_pattern,
```

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -58,8 +79,18 @@ def connect(self, port):
 
         self._connect(port, baudrate=38400)
```
You are not changing this, but ideally this baudrate should also come from the config.

wtgee commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -58,8 +79,18 @@ def connect(self, port):
+        # Send get position command to see if response ends with '#', proper of the astromechs
+        res = self._send_command("P#")
+
+        if res and res.endswith("#"):
+            try:
+                dev = usb.core.find(idVendor=0x0403, idProduct=0x6001)
+                self._serial_number = usb.util.get_string(dev, dev.iSerialNumber)
+                # Return string that makes it clear there is no serial number
+                return self._serial_number
+
+            except Exception:
+                raise error.PanError(f"{self.name} not found in {self.port}")
```
I still think the exception should be raised when the `if` fails, not when the serial number check fails.  And can we check it for a valid position response rather than the generic `#`?

wtgee replied:
```suggestion
                raise error.BadSerialConnection(f"{self.name} not found in {self.port}")
```

wtgee commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -64,57 +63,18 @@ def __init__(self,
                  initial_position=None,
                  dev_node_pattern='/dev/tty.USA49*.?',
                  max_command_retries=5,
+                 serial_number_pattern=serial_number_pattern,
```
Similar to the astromechanics, the `dev_node_pattern` should default to `None` and these parameters probably don't need to be specified.

wtgee commented on src/panoptes/pocs/focuser/serial.py:
```diff
@@ -15,13 +16,55 @@ class AbstractSerialFocuser(AbstractFocuser):
     _assigned_nodes = []
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, dev_node_pattern="", initial_position=None,
+                 serial_number_pattern="", *args, **kwargs):
```
Default to `None` for unknown things.

```suggestion
    def __init__(self, dev_node_pattern=None, initial_position=None,
                 serial_number_pattern=None, *args, **kwargs):
```

wtgee replied:
For the `serial_number_pattern`, I would create your own `re` object if given a string.

wtgee commented on src/panoptes/pocs/focuser/serial.py:
```diff
@@ -15,13 +16,55 @@ class AbstractSerialFocuser(AbstractFocuser):
+        if serial_number_pattern.match(self.port):
+            # Have been given a serial number
+            self.logger.debug('Looking for {} ({})...'.format(self.name, self.port))
+
+            if AbstractSerialFocuser._adaptor_nodes is None:
+                # No cached device nodes scanning results, need to scan.
+                self.logger.debug('Getting serial numbers for all connected focusers')
+                AbstractSerialFocuser._adaptor_nodes = {}
+                # Find nodes matching pattern
+                device_nodes = glob.glob(dev_node_pattern)
+
+                # Open each device node and see if a focuser answers
+                for device_node in device_nodes:
+                    try:
+                        serial_number = self.connect(device_node)
+                        AbstractSerialFocuser._adaptor_nodes[serial_number] = device_node
+                    except (serial.SerialException, serial.SerialTimeoutException, AssertionError):
+                        # No focuser on this node.
+                        pass
+                    finally:
+                        self._serial_port.close()
+
+                if not AbstractSerialFocuser._adaptor_nodes:
+                    message = 'No focuser devices found!'
+                    self.logger.error(message)
+                    warn(message)
+                    return
+                else:
+                    self.logger.debug('Connected focusers: {}'.format(AbstractSerialFocuser._adaptor_nodes))
+
+            # Search in cached device node scanning results for serial number
+            try:
+                device_node = AbstractSerialFocuser._adaptor_nodes[self.port]
+            except KeyError:
+                message = 'Could not find {} ({})!'.format(self.name, self.port)
+                self.logger.error(message)
+                warn(message)
+                return
+            self.logger.debug('Found {} ({}) on {}'.format(self.name, self.port, device_node))
+            self.port = device_node
```
This should all be moved out of the class instance and become a class method that is called before object initialization.

See discussion in PR review.

wtgee reviewed (changes requested):
I realize this is just a copy/paste job from what was there, but I think if this is not a huge priority time wise then we should take the time to clean up the constructor.

Ideally the `dev_node_pattern` does not happen in an object instance so I'm proposing to make this a class method that would be called prior to object initialization. 

I don't see any advantage to this happening dynamically upon creation, but let me know if there is a valid reason still.  @AnthonyHorton @JAAlvarado-Montes

JAAlvarado-Montes commented:
Hi @wtgee @AnthonyHorton @danjampro ,

I have addressed all of @wtgee's comments. I'll submit a different PR to `huntsman-config` to add  `baudrate` (birger and astromech); `id_vendor` and `id_product` (only for astromech). The classes are now much more organized and I even find the code more readable.

Everything was tested on Huntsman's lens Tuc with different Astromechanics focusers. Please let me know any suggestion or comment.

PS. Tests keep failing because of the new `pyusb` module that astromechanics.py is using. However, I've added it as a new requirement for POCS as @wtgee indicated, so I'm not sure what is happening.

Cheers,
Jaime A.

danjampro commented on src/panoptes/pocs/focuser/astromechanics.py:
```diff
@@ -8,25 +12,25 @@ class Focuser(AbstractSerialFocuser):
+    def __init__(self, name='Astromechanics Focuser', model='Canon EF-232', id_vendor=None,
+                 id_product=None, *args, **kwargs):
+
+        self._id_vendor = id_vendor
```
Minor detail, but shouldn't these be `_vendor_id` and `_product_id`?

danjampro commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -47,77 +41,23 @@ class Focuser(AbstractSerialFocuser):
-
-    Additional positonal and keyword arguments are passed to the base class, AbstractFocuser. See
-    that class' documentation for a complete list.
+        max_command_retries (int, optional): Maximum number of times to retry a command. Default 5
```
Do we need a `max_retries` for these focusers?

JAAlvarado-Montes replied:
Yes we do, it is used as the maximum number of times to retry a command when it isn't properly sent

danjampro reviewed (commented):
Hi @JAAlvarado-Montes looks good. I'm not an expert in the focuser code so I'll leave it to the others to give the final review. The test coverage is showing a bit of a drop, can anything be done about this?

lspitler commented:
@JAAlvarado-Montes what is the status of this? Do you need a review from @wtgee ?

JAAlvarado-Montes commented:
@lspitler, as far as I'm concerned all the changes from this PR are ready and working with the Astromechs. I followed all of @wtgee's and @danjampro's comments. Also, tests are passing with no issues.

lspitler commented:
@wtgee @AnthonyHorton we're about to deploy the AstroMechs in ~7 days. @JAAlvarado-Montes is this PR required for deploying them?

wtgee reviewed (commented):
@JAAlvarado-Montes I did this review a long time ago but apparently never hit submit, sorry about that!  I think most of the comments still apply but let me know if something doesn't make sense.

JAAlvarado-Montes commented:
Replaced by #1125
